### ISSUE TITLE
Unable to reconnect to chromedriver when switching back from native to webview 

### ISSUE THREAD

**Issue by glorylo** [appium/appium#4502]
Using fairly recent appium on android hybrid app.  You can switch to webview but if you return back to native and then try to switch back to webview, it timesout as it is unable to reconnect to the chromedriver.

When reconnecting...

```
 info: [debug] Connecting to chrome-backed webview
> info: [debug] Found existing Chromedriver for context 'WEBVIEW_com.someapp. Using it.
> info: [debug] Making http request with opts: {"url":"http://127.0.0.1:9515/wd/hub/session/3a3a448ad0ab32fe889d61f4b0dc6cd6/url","method":"GET"}
> info: [debug] Didn't get a new command in 60 secs, shutting down...
```

I am not sure it is because the existing session and its colliding.

**Comment by orestesd** [comment]
Hi glorylo.

Did you manage to solve it? 
I have the same problem, I think is related to chromedriver session expiration

https://discuss.appium.io/t/reusing-existent-chromedriver-session-gets-hang-appium-1-3-5/3093

**Comment by moizjv** [comment]
Looks like an issue

**Comment by glorylo** [comment]
@orestesd I never solved this issue :(

**Comment by orestesd** [comment]
Will it fix the problem?

https://github.com/appium/appium/releases/tag/v1.3.7
> fix failure to remap session id in proxied responses

**Comment by SealviewTW** [comment]
I am using server version 1.3.7.2 on Windows and I still have this problem.
Didn't try it on 1.4.0 since my code is not working anymore with 1.4.0, and I do not wish to change the whole project because of server update.

**Comment by noymn** [comment]
Do you have any estimation about when will this be fixed? I tried several workarounds but none worked... I don't know if changing the session id for the file `lib/devices/android/android-hybrid.js``

When the switch is requested appium hangs here: 
```
 // check the status by sending a simple window-based command to ChromeDriver
    // if there is an error, we want to recreate the ChromeDriver session
    var url = '/wd/hub/session/' + this.proxySessionId + '/url';
    this.chromedriver.proxyTo(url, 'GET', {}, function (err, res, body) {
      body = JSON.parse(body);
```

**Comment by inertjade** [comment]
I face this problem on my Ubuntu with Appium version 1.4.0. However, this problem doesn't show up on my Mac running Appium 1.4.0. This is what I see when it tries to reuse an old session of ChromeDriver.

```
info: [debug] Found existing Chromedriver for context 'WEBVIEW_1'. Using it.
info: Proxying [GET /url] to [GET http://127.0.0.1:9515/wd/hub/session/eaf2f9c9e5d20d285d7d92cbe8eb6fa5/url] with no body
info: Got response with status 200: {"sessionId":"eaf2f9c9e5d20d285d7d92cbe8eb6fa5","status":100,"value":{"message":"chrome not reachable\n  (Session info: chrome=41.379.2250.2)\n  (Driver info: chromedriver=2.15.322455 (ae8db840dac8d0c...
info: [debug] ChromeDriver is not associated with a window. Re-initializing the session.
info: Restarting chromedriver
```

Not sure why the Appium client on Ubuntu doesn't restart chromedriver.

**Comment by moizjv** [comment]
Can you try on 1.4.6?

**Comment by esoests** [comment]
I am using 1.4.6 and 1.4.10 on mac with real devices, the issue still exists.

 ```java
info: [debug] Getting a list of available webviews
info: [debug] executing cmd: /Applications/android-sdk-macosx/platform-tools/adb -s 4d0042cddc8e611b shell "cat /proc/net/unix"
info: [debug] WEBVIEW_12090 mapped to pid 12090
info: [debug] Getting process name for webview
info: [debug] executing cmd: /Applications/android-sdk-macosx/platform-tools/adb -s 4d0042cddc8e611b shell "ps"
info: [debug] Parsed pid: 12090 pkg: com.app
info: [debug] from: u0_a235,12090,2361,2441620,334736,ffffffff,00000000,R,com.app
info: [debug] returning process name: com.app
info: [debug] Available contexts: NATIVE_APP,WEBVIEW_com.app
info: [debug] ["WEBVIEW_com.amazon.mShop.android"]
info: [debug] Available contexts: NATIVE_APP,WEBVIEW_com.app
info: [debug] Connecting to chrome-backed webview
info: [debug] Found existing Chromedriver for context 'WEBVIEW_com.app'. Using it.
info: JSONWP Proxy: Proxying [GET /url] to [GET http://127.0.0.1:9515/wd/hub/session/2a71f47313c9b97239ca23d02d8c739b/url] with no body

**Comment by moizjv** [comment]
@esoests We are very soon releasing 1.5 and we have added a cap `recreateChromeDriverSessions` that will not reuse old webviews that would fix this issue

**Comment by esoests** [comment]
Looking forward to use it, thanks!

**Comment by esoests** [comment]
Looking forward to use it, thanks.

**Comment by NehaGoyalFK** [comment]
Hi Moiz,

I am facing the same issue -
info: [debug] Available contexts: NATIVE_APP,WEBVIEW_com.flipkart.android
info: [debug] ["WEBVIEW_com.flipkart.android"]
info: [debug] Available contexts: NATIVE_APP,WEBVIEW_com.flipkart.android
info: [debug] Connecting to chrome-backed webview
info: [debug] Found existing Chromedriver for context 'WEBVIEW_com.flipkart.android'. Using it.
info: JSONWP Proxy: Proxying [GET /url] to [GET http://127.0.0.1:9515/wd/hub/session/e6bfd7c7772d13df3bbf8fa449143298/url] with no body
info: [debug] Didn't get a new command in 60 secs, shutting down...
info: Shutting down appium session

Please let me know when will this fix be available as the app I work on heavily uses both the contexts and switching between them is hampering my execution cycles.

**Comment by NehaGoyalFK** [comment]
@moizjv 
I have also tried the beta version of 1.5 and it has not yet resolved the issue. I am hitting the same issue. Please have a look.

**Comment by NehaGoyalFK** [comment]
@esoests @noymn @glorylo 
Have any of you been able to resolve this issue?
I have been facing this issue, and tried multiple things
1)Download appium 1.5 beta
2)Upgrade chromedriver to 2.20
3)Check if there were multiple handles which were screwing the driver
But none of the above is helping me.

**Comment by mi2pankaj** [comment]
Hello everyone!! I could get a fix by monitoring chromedriver all the time during test execution, chromedriver becomes unresponsive while switching back to webview from natiove_app, I developed a small piece of code available at https://github.com/mi2pankaj/chromeDriverHandler, it has solved my all problems, just add this jar chromedriverHandler.jar  to your project and use ChromedriverHandler.chromeDriverHandlerThread().start(); in your @beforeclass / @beforetest and ChromedriverHandler.chromeDriverHandlerThread().stop(); in your @afterclass / @aftertest.

**Comment by moizjv** [comment]
All can some one paste appium logs from appium 1.5 run. Please use `recreateChromeDriverSessions` capability to `true`

**Comment by moizjv** [comment]
I have added this cap for the above issue, which will not reuse chromedriver session instead recreate chrome driver each time

**Comment by mi2pankaj** [comment]
@moizjv : recreateChromeDriverSessions is a new capability introduced in 1.5 ?? 
Just to confirm: using this, appium restarts the chromedriver when doesn't get response from chromedriver ?? Because I also developed small piece of code to do that in chromedriverhandler repo.

**Comment by drew-royster** [comment]
I added this line of code to my project as you explained

_"capabilities.setCapability("recreateChromeDriverSessions", true)"_

but i'm still getting the same error...  

_info: [debug] Found existing Chromedriver for context 'WEBVIEW_com.netiq.mobileaccessforandroid'. Using it._

Is there a way to manually delete the chromedriver after I use it the first time so that the second time I call for a chromedriver it builds a new one?

**Comment by drew-royster** [comment]
I added this line of code to my project as you explained

"capabilities.setCapability("recreateChromeDriverSessions", true)"

but i'm still getting the same error...

info: [debug] Found existing Chromedriver for context 'WEBVIEW_...'. Using it.

Is there a way to manually delete the chromedriver after I use it the first time so that the second time I call for a chromedriver it builds a new one?

**Comment by mi2pankaj** [comment]
Can you try this as mentioned in above conversation: piece of code available at https://github.com/mi2pankaj/chromeDriverHandler, add this jar chromedriverHandler.jar to your project and use ChromedriverHandler.chromeDriverHandlerThread().start(); in your @beforeclass / @beforetest and ChromedriverHandler.chromeDriverHandlerThread().stop(); in your @afterclass / @aftertest.

**Comment by dimitarz** [comment]
`capabilities.setCapability("recreateChromeDriverSessions",` `true);`

[debug] [AndroidDriver] recreateChromeDriverSessions set to true; killing existing chromedrivers

Worked like a charm - thanks!

**Comment by moizjv** [comment]
@mi2pankaj  yeah you can use the new cap instead of workaround code :)

**Comment by KrazySush** [comment]
Hi All,

I am automating hybrid mobile app.
For Android 6.0 & above, when switching to Webview, execution halts for sometime (say 5-10 mins).

I have following configurations:
Mac OS X El Capitan v 10.11.5
Appium 1.5.2

I have tried adding following cap:
**capabilities.setCapability("recreateChromeDriverSessions", true);**

But still facing same issue.

Need help

**Comment by talarari** [comment]
i tried using recreateChromeDriverSessions  but still doesnt help in my case
i've created a reproducible sample here:
https://github.com/talarari/AppiumSwitchContextBug

would appreciate any help you can offer, willing to help with this not sure how

**Comment by pradyumna04** [comment]
I too faced this issue, resolved by using the latest chromedriver build 2.23 and AndroidMobileCapabilityType.RECREATE_CHROME_DRIVER_SESSIONS capability.

This helped: https://github.com/appium/appium/issues/6634

**Comment by jayeshust** [comment]
Hi I am trying to run tests on a real Android device with Appium in Ubuntu. I can see the app is installing and launching but when switching to web-view I am getting the below issue. Please help.

 [debug] [AndroidDriver] Found webviews: ["WEBVIEW_com.openstream.cueme.services.workbench"]
    [debug] [AndroidDriver] Available contexts: ["NATIVE_APP","WEBVIEW_com.openstream.cueme.services.workbench"]
    [debug] [AndroidDriver] Connecting to chrome-backed webview context 'WEBVIEW_com.openstream.cueme.services.workbench'
    [debug] [AndroidDriver] A port was not given, using random port: 8000
    [debug] [Chromedriver] Changed state to 'starting'
    [Chromedriver] Set chromedriver binary as: /usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64
    [Chromedriver] Killing any old chromedrivers, running: pkill -15 -f "/usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64.*--port=8000"
    [Chromedriver] No old chromedrivers seemed to exist
    [Chromedriver] Spawning chromedriver with: /usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64 --url-base=wd/hub --port=8000 --adb-port=5037
    [Chromedriver] [STDERR] /usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64: 1: /usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64:
    [Chromedriver] [STDERR] ELF: not found
    [Chromedriver] [STDERR] /usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64: 2: /usr/local/lib/node_modules/appium/node_modules/appium-chromedriver/chromedriver/linux/chromedriver_64: Syntax error: "(" unexpected
    [Chromedriver] Chromedriver exited unexpectedly with code 2, signal null
    [debug] [Chromedriver] Changed state to 'stopped'
